feat: 카카오 소셜 로그인 기능 추가(#3)

handseokjin opened the pull request:
- [x] 🔀 PR 제목의 형식을 잘 작성했나요? e.g. `[feat] PR을 등록한다.` 
- [x] 💯 테스트는 잘 통과했나요?
- [x] 🏗️ 빌드는 성공했나요?
- [x] 🧹 불필요한 코드는 제거했나요?
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?
- [x] 💻 git rebase를 사용했나요?

## 작업 내용
- 카카오 로그인 기능 전반 구현
  - 로그인 하려는 계정의 이메일 및 닉네임 정보 가져옴
  - WebClient 라이브러리를 통한 카카오 서버와의 통신 기능 추가
  - intellj 내 카카오 관련 환경설정 값 추가
- 로그인 성공 및 실패 시 응답하는 부분 수정
  - 기존 : 문자열 → 수정 : 객체(jwt 외 닉네임도 반환하도록 추가)
- utils 클래스 함수 static으로 변경 및 BasicUtils 문자열을 객체로 바꾸는 함수 추가
 ※ 저번에 수정해서 머지했는데 이관하면서 반영이 안 되었던 것 같습니다. 그래서 다시 수정했습니다.

## 스크린샷
1. 카카오 로그인 기능 관련(리액트로 카카오 로그인 프론트 부분 간단히 만들어 본 후 서버와 테스트 해 보았습니다.)
 - 최초 로그인
  1). 최초 로그인 시 아이디 및 비밀번호 입력 템플릿 출력
<img width="1271" alt="카카오 로그인 버튼" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/df94fdcc-2d5b-473c-9a09-168afb52c7e4">
<img width="1255" alt="최초 로그인 시 1" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/2462b2a1-d650-4561-a555-df3de17ff1c9">
  2). 입력한 아이디가 db에 없으면 401 에러 발생(현명님께서는 저 때 회원가입 창으로 리다이렉트 하도록 회의 하여 전달했습니다)<br/>
<img width="1276" alt="최초 로그인 시 2(회원가입 안 되어 있어서 401 발생)" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/e149e91f-5317-4267-ac81-4f332bb96eed">

 - 소셜 로그인으로 회원가입 완료된 후 로그인
   1). 동일하게 아이디 및 비밀번호 입력 템플릿 출력
<img width="1255" alt="최초 로그인 시 1" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/2462b2a1-d650-4561-a555-df3de17ff1c9">
   2). db에 해당 이메일이 저장되어 있는 경우 홈 화면으로 리다이렉트 <br/>
    - 제가 간단히 만든 리액트에서는 db에 해당 이메일이 저장되어 있어 로그인이 성공했을 경우(200번 응답코드 반환) 저렇게 MainPage 라는 글자가 출력되는 MainPage로 이동하도록 만들었습니다(현명님께서는 로그인 성공 시 특정 페이지로 리다이렉트 하거나 어떻게 할 것 같습니다.)
<img width="1278" alt="추후 로그인(로그인 성공 시 메인 페이지로 이동)" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/0007357d-f7bb-4715-87bf-798173809ddb">
    - 정상 로그인 시 jwt 생성 및 반환
<img width="768" alt="추후 로그인(정상적으로 jwt 생성)" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/32a0c114-afda-4d90-a7bc-567cc142f488"> <br/>
    - 로그인 한 이메일 및 닉네임 정상적으로 가져옴 <br/>
<img width="417" alt="추후 로그인(이메일 닉네임 가져옴)" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/f7fe56c2-3f23-4684-8e3e-314d12bc8f5f">

<br/><br/>

2. 응답하는 부분 객체로 수정
  - 기존 회원가입 정상 동작
<img width="846" alt="기존 회원가입 정상 완료" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/2f7ec1e5-6787-4e36-bd80-de0ca8316aea">
<br/>
  - 응답 코드 객체로 변경 및 닉네임 추가
<img width="860" alt="로그인 수정 정상 작동" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/f89e809c-1126-417c-8955-2354047b6936">


## 주의사항

Closes #3

Commit:
feat: 카카오 소셜 로그인 기능 추가(#3)

handseokjin commented:
기존 로그인이랑 최대한 중복 없게 하고, 카카오 로그인 관련 컨트롤러 부분은 간단하게 하기 위해 좀 오래 걸렸습니다.
캡처를 최대한 자세히 한다고 했는데 혹시 부족한 부분 있으면 수정하겠습니다!

gyujin-Kimm commented:
코드를 전체적으로 봤는데 인증/인가 부분은 역시 어렵네요😂
석진님이 많이 고민하신게 보였습니다!👍
개인적으로 리액트까지 간단하게 구현하여 테스트해보신게 정말 대단하십니다! 저도 많이 배우겠습니다!

acisliver commented on src/main/java/com/nakaligoba/backend/utils/BasicUtils.java:
유틸 클래스면 static 메서드로 구현하고 컴포넌트는 빼는 게 좋을 것 같습니다!

handseokjin replied:
음, 혹시 몰라서 빼둔 건데 그러면 메소드 3개 모두 사용하는 곳으로 private 함수로 해서 옮겨두겠습니다!

acisliver commented on src/main/java/com/nakaligoba/backend/controller/MemberController.java:
```diff
@@ -34,6 +39,17 @@ public ResponseEntity<SignupResponse> signup(@Valid @RequestBody SignupRequest r
+    public ResponseEntity<SigninResponse> kakaoSignin(@Valid @RequestBody KakaoSigninRequest request) {
+        SigninResponse signinResponse = memberService.kakaoSignin(request.getAuthCode());
+
+        if (!"".equals(signinResponse.accessToken)) {
```
```suggestion
        if (StringUtils.hasText(signinResponse.accessToken)) {
```

이렇게 작성하는 게 더 가독성이 높을 것 같네요

handseokjin replied:
아 맞네요. 말씀하신 것처럼 하는 게 훨씬 가독성 높아 보입니다!

acisliver commented on src/main/java/com/nakaligoba/backend/controller/MemberController.java:
```diff
@@ -133,9 +149,37 @@ static class SigninRequest {
+
+    @Data
+    @NoArgsConstructor
+    static class KakaoSigninRequest {
+        @NotBlank
+        private String authCode;
+    }
```
controller 하위에 dto를 만들기로 했으니 다음에 옮기죠!

handseokjin replied:
넵 알겠습니다!

acisliver commented on src/main/java/com/nakaligoba/backend/service/MemberService.java:
```diff
@@ -49,6 +56,42 @@ public void signup(MemberDto memberDto) {
+        }
+    }
+
+    public String getMemberJwt(String email) {
```
마찬가지로 public일 이유가 있나요?

handseokjin replied:
그렇..네요. 제가 너무 급하게 했나 봅니다;;

handseokjin commented on src/main/java/com/nakaligoba/backend/utils/JwtUtils.java:
```diff
@@ -7,7 +7,7 @@
 @Component
```
아 맞네요. 전에 머지했을 때는 다 뺐었는데 깜빡 했네요;; 객체 생성한 것도 다 지워야 할 것 같네요

handseokjin commented on src/main/java/com/nakaligoba/backend/service/MemberService.java:
```diff
@@ -49,6 +56,42 @@ public void signup(MemberDto memberDto) {
         memberRepository.save(memberEntity);
     }
 
+    public KakaoSigninTokenResponse getKakaoSigninToken(String kakaoAuthCode) {
+        return kakaoWebClient.getKakaoSigninToken(kakaoAuthCode);
+    }
+
+    public KakaoSigninUserInfoResponse getKakaoSigninUserInfo(String accessToken) {
+        return kakaoWebClient.getKakaoSigninUserInfo(accessToken);
+    }
```
그렇..네요. 제가 너무 급하게 했나 봅니다;;

handseokjin commented:
대은님 규진님 리뷰 감사합니다. 말씀하신 부분들 수정해서 다시 커밋해 두겠습니다. 급하게 하다 보니 많이 놓쳤네요. 다음에 좀 더 꼼꼼히 보고 올리겠습니다!

Commit:
chore: 리뷰를 통해 언급받은 부분의 수정에 대한 커밋

1. JwtUtils 내 @Component 어노테이션 제거
2. BasicUtils 제거(해당 클래스 내 메소드는 호출되는 메소드 내 private으로 선언)
3. StringUtils.hasText로 수정 적용
4. 언급해주신 메소드 접근 제어지시자 public에서 private으로 수정
 - getKakaoSigninToken(), getKakaoSigninUserInfo(), getMemberJwt()

※ 모두 수정 완료 후, 테스트 결과 정상 작동 확인했습니.

handseokjin commented:
말씀해 주신 부분들 전부 수정 반영해서 테스트 한 후에 추가 커밋해 두었고, 머지해 두겠습니다!
